Better error handling

mkangia opened the pull request:
https://manage.dimagi.com/default.asp?276927

Show error message to user 
```
Looks like you are assigning a location name/code to a different location in the same request. 
Please do this in two separate updates by using a temporary name to free up the name/code to be re-assigned.
```

if they are trying to swap names/codes for two locations or trying to use name/code for a location to another

Doing such an update in a single request results in IntegrityError at database due to unique constraint while the updates are being done sequentially.

fyi @emord @esoergel

mkangia committed:
Show error message to user if they are trying to swap names/codes for two locations or trying to use name/code for a location to another
Doing such an update in a single request results in IntegrityError at database due to unique constraint while the updates are being done sequentially.

esoergel commented on corehq/apps/locations/views.py:
```diff
@@ -342,14 +344,33 @@ def mk_loctype(name, parent_type, administrative, has_user,
+                        "Looks like you are assigning a location name/code to a different location "
+                        "in the same request. Please do this in two separate updates by using a "
+                        "temporary name to free up the name/code to be re-assigned."))
+                    )
```
I think we should be able to work around this in some way - would wrapping all the saves in a single transaction allow this to go through as-is?

mkangia replied:
@esoergel i don't think that would help since the save would still happen one by one and the unique clause would fail?

esoergel replied:
Dang, I was hoping postgres would defer unique checking until the transaction is finished, but that doesn't appear to be the case (just tried it out in a shell session).

Another option is to assign one of the location types a temporary name (say, a UUID), then make a second pass and save it again with the real name.  It's an extra round-trip to the db, but we expect there to be a very small number of location types on any given domain, and it beats making the user do that manually.

esoergel replied:
Huh, okay.  I do think that would be better behavior, but I haven't tried it to see how complex that is to implement.  I'll leave it up to your judgement on whether it's worth it.

snopoke replied:
I had similar thoughts as @esoergel but I guess if it does introduce a lot of complexity then fine to leave as is for now. If this comes up again we can do it then.

mkangia replied:
Thank you.

mkangia commented:
@esoergel 
any suggestions on the stickler error i am getting here. Can it be ignored?

esoergel commented:
@mkangia 
> W1656 dict.values referenced when not iterating

That's a python 2-to-3 issue - in Python 3, `.values()` returns an iterator, which can break some things.  The fix is to immediately wrap the call with `list`

```python
names = list(payload_loc_type_name_by_pk.values())
```

mkangia commented:
Cool, thanks @esoergel

mkangia committed:
add tests for post view

mkangia committed:
lint

Call list immediately becuase in python 3 values() return an iterator which can cause issues

mkangia commented:
Could not find a setup to test views so added a new one and added some tests.
@esoergel

mkangia commented:
@dimagi/product this should be happening very often but just fyi in case you have any feedback/suggestion.

mkangia commented:
Offlined with @djmore9 and this is okay.

snopoke commented:
:+1: still has the 'do not merge' label

mkangia commented:
@snopoke 
removed the label and merged.
Thank you!
